doc/developer/design/20230714_optimizer_interface.md

```diff
@@ -0,0 +1,336 @@
+   without major refactoring of the ADAPTER call sites.
+3. Be aligned with the Platform V2 goals.
+
+## Out of Scope
+
+<!--
+What does a solution to this problem not need to address in order to be
+successful?
+
+It's important to be clear about what parts of a problem we won't be solving
+and why. This leads to crisper designs, and it aids in focusing the reviewer.
+-->
+
+1. Immediately solve long-standing issues with the optimizer pipeline. While the
+   design should accomodate natural mechanisms for solving those, we don't want
+   to implement everything within the initial migration effort. For example: an
+   optimizer might want to provide shared context (for typing invariants, for
+   Prometheus metrics, for feature flags) across all of its phases. While
+   introdcing the notion of a shared context is part of the scope, threading it
+   through is is not.
+2. Consolidate the various stages of the optimization pipeline into a single
+   stage. Again, while this can be a long-term goal, it seems too much of an ask
+   to do this refactor when we introduce the new optimizer interface. Rather, we
+   can revisit the (hopefully now cleaner) code when we are done and design a
+   path towards merging adjacent stages together whenever possible.
+3. The new API will also allow us to have a clearer notion of "optimizer
+   version", which could be a useful tool in migration. For example, we can do
+   A/B migration testing: when we are considering a change to the optimizer, we
+   can optimize all queries with both — identifying and anonymously logging
+   production queries that will behave differently in the new optimizer. Again,
+   this is something that can be facilitated by the proposed changed, but will
+   not be scoped as part of them.
+
+## Solution Proposal
+
+<!--
+What is your preferred solution, and why have you chosen it over the
+alternatives? Start this section with a brief, high-level summary.
+
+This is your opportunity to clearly communicate your chosen design. For any
+design document, the appropriate level of technical details depends both on
+the target reviewers and the nature of the design that is being proposed.
+A good rule of thumb is that you should strive for the minimum level of
+detail that fully communicates the proposal to your reviewers. If you're
+unsure, reach out to your manager for help.
+
+Remember to document any dependencies that may need to break or change as a
+result of this work.
+-->
+
+At the high level, we want to have a minimalistic API defined by the following
+trait:
+
+```rust
+pub trait Optimize<'ctx, From>: Send + Sync
+where
+    From: Send + Sync,
+{
+    type To: Send + Sync + 'ctx;
+
+    /// Execute the optimization stage, transforming the input plan of type
+    /// `From` to an output plan of type [`Optimize::To`].
+    fn optimize<'s: 'ctx>(&'s mut self, plan: From) -> Result<Self::To, OptimizerError>;
+}
+```
```

Comment:
What's the purpose of the `'cix` lifetime?

Reply:
Added a comment on top of the trait in the latest revision with an explanation.